### PR TITLE
[REVIEW] IO Readers: Support memory buffer, file-like object, and URL inputs

### LINKED ISSUE #2043: [FEA] Add memory buffer input support to ORC/Parquet readers
ORC/Parquet readers currently only accept file names as input.  We need support for memory buffer input where the bytes of a complete ORC/Parquet file are passed in as buffer. 

Note:  There is no need to support processing of byte ranges.

### LINKED ISSUE #2062: [FEA] cudf.read_csv and other readers should support reading from a URL
With pandas, I can read a CSV (or other filetypes) directly from a URL as opposed to a file on disk.

```
import pandas as pd

url="https://raw.githubusercontent.com/cs109/2014_data/master/countries.csv"
c=pd.read_csv(url)
```

It would be nice to be able to do the same thing in cudf instead of having to first download the file or read it into a StringIO object.

### PR DESCRIPTION
**Pull request by j-ieong**
Fixes #2043 and fixes #2062.

* Add support for reading from in-memory buffers, such as `ByteIO` to ORC and Parquet readers
    - use `memoryviews` in Cython for direct data access without copy
* Add support for reading from URL (such as `http`) to all readers
* Update Python documentation for functions to describe all supported inputs
* Add pytests

### CODE REVIEW

**Comment by j-ieong** [comment]
rerun tests

File: python/cudf/tests/test_orc.py [hunk 1]
```diff
@@ -77,6 +108,26 @@ def test_orc_reader_filenotfound(tmpdir):
         cudf.read_orc(tmpdir.mkdir("cudf_orc"))
 
 
+def test_orc_reader_local_filepath():
+    path = '~/TestOrcFile.test1.orc'
```
Comment (kkraus14) [inline python/cudf/tests/test_orc.py]: This doesn't look right, should this be datadir similar to other tests?

File: python/cudf/tests/test_parquet.py [hunk 2]
```diff
@@ -260,6 +290,23 @@ def test_parquet_reader_filenotfound(tmpdir):
         cudf.read_parquet(tmpdir.mkdir("cudf_parquet"))
 
 
+def test_parquet_reader_local_filepath():
+    fname = '~/spark_timestamp.snappy.parquet'
```
Comment (kkraus14) [inline python/cudf/tests/test_parquet.py]: This one too.

File: python/cudf/tests/test_orc.py (hunk 1 above)
Comment (j-ieong) [reply python/cudf/tests/test_orc.py]: PANDAS' allows giving a filepath of `~/xxx.csv` and it'll auto-expand to `/home/jieong/xxx.csv`. 

This PR adds that user home path expansion functionality but not sure how to test in CI. I would need to write a temporary file to the CI machine's user home directory and read it to actually exercise it...is that possible? Right now, this test is skipped in CI. Or any other idea?

Comment (kkraus14) [reply python/cudf/tests/test_orc.py]: No, `datadir` will point to `python/cudf/tests/data` which is where I would expect these files to live. Is that not correct?

Comment (j-ieong) [reply python/cudf/tests/test_orc.py]: Hrmmm, so `~` doesn't contain part of `datadir` on the CI machines?

PANDAS expands a file path of `~/xxx.csv` to `/home/jieong/xxx.csv`. This PR adds that functionality too, so I want a test case for it...

Comment (kkraus14) [reply python/cudf/tests/test_orc.py]: Is this specifically for testing that the `~` character is handled in filepaths passed down to the C++ layer?

Comment (j-ieong) [reply python/cudf/tests/test_orc.py]: The expansion is done in the Python layer currently, in `get_filepath_or_buffer()` which is similar to where PANDAS does it.

Comment (kkraus14) [reply python/cudf/tests/test_orc.py]: Alternative workarounds here aren't very good so I guess lets just leave this as is. Do we want to change the filename to something more generic for both here and the Parquet file? Below it's named `spark_timestamp.snappy.parquet`

Comment (j-ieong) [reply python/cudf/tests/test_orc.py]: Done!

File: python/cudf/tests/test_parquet.py (hunk 2 above)
Comment (j-ieong) [reply python/cudf/tests/test_parquet.py]: Resolve to same as above.

**Pull request merged**